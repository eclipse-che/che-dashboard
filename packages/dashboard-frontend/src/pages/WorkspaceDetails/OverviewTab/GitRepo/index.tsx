/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Button, FormGroup } from '@patternfly/react-core';
import { CopyIcon } from '@patternfly/react-icons';
import { load } from 'js-yaml';
import React from 'react';
import CopyToClipboard from 'react-copy-to-clipboard';

import { CheTooltip } from '@/components/CheTooltip';
import overviewStyles from '@/pages/WorkspaceDetails/OverviewTab/index.module.css';
import devfileApi from '@/services/devfileApi';
import { Workspace } from '@/services/workspace-adapter';
import {
  DEVWORKSPACE_DEVFILE_SOURCE,
  DEVWORKSPACE_METADATA_ANNOTATION,
} from '@/services/workspace-client/devworkspace/devWorkspaceClient';

export type Props = {
  workspace: Workspace;
};

export type State = {
  timerId: number | undefined;
};

class GitRepoFormGroup extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      timerId: undefined,
    };
  }

  private handleCopyToClipboard(): void {
    let { timerId } = this.state;
    if (timerId !== undefined) {
      window.clearTimeout(timerId);
    }
    timerId = window.setTimeout(() => {
      this.setState({
        timerId: undefined,
      });
    }, 3000);
    this.setState({ timerId });
  }

  public getSource(workspace: devfileApi.DevWorkspace): {
    isUrl: boolean;
    gitRepo: string;
    fieldName: string;
  } {
    const source = {
      isUrl: false,
      gitRepo: '',
      fieldName: '',
    };

    let devfileSourseStr =
      workspace.metadata.annotations?.[DEVWORKSPACE_METADATA_ANNOTATION]?.[
        DEVWORKSPACE_DEVFILE_SOURCE
      ];

    if (devfileSourseStr === undefined) {
      devfileSourseStr =
        workspace.spec?.template?.attributes?.[DEVWORKSPACE_METADATA_ANNOTATION]?.[
          DEVWORKSPACE_DEVFILE_SOURCE
        ];
    }

    if (devfileSourseStr === undefined) {
      return source;
    }

    const devfileSourse = load(devfileSourseStr) as {
      scm?: {
        repo?: string;
        fileName?: string;
      };
      factory?: {
        params?: string;
      };
    };

    const factoryParams = devfileSourse?.factory?.params;
    if (factoryParams === undefined) {
      return source;
    }

    const paramsArr = factoryParams.split('&');
    if (paramsArr.length === 0) {
      return source;
    }

    paramsArr.forEach(param => {
      const [key, value] = param.split('=');
      if (key === 'url') {
        if (!source.gitRepo && paramsArr.length === 1) {
          source.gitRepo = value;
        } else {
          source.gitRepo = value + '?' + source.gitRepo;
        }
        source.isUrl = new RegExp('^http[s]?://').test(value);
        source.fieldName = source.isUrl ? new URL(value).pathname.replace(/^\//, '') : value;
        if (source.fieldName.length > 50) {
          source.fieldName = source.fieldName.substring(0, 50) + '...';
        }
      } else {
        if (source.gitRepo.length !== 0 && !source.gitRepo.endsWith('?')) {
          source.gitRepo = source.gitRepo + '&';
        }
        source.gitRepo = source.gitRepo + key + '=' + value;
      }
    });

    return source;
  }

  public render(): React.ReactNode {
    const { workspace } = this.props;
    const { timerId } = this.state;

    const { gitRepo, fieldName, isUrl } = this.getSource(workspace.ref);
    if (!gitRepo || !fieldName) {
      return <></>;
    }

    return (
      <FormGroup label="Git repo URL">
        {isUrl ? (
          <Button component="a" variant="link" href={gitRepo} target="_blank" isInline>
            {fieldName}
          </Button>
        ) : (
          <span className={overviewStyles.readonly}>{fieldName}</span>
        )}
        <CheTooltip content={timerId ? 'Copied!' : 'Copy to clipboard'}>
          <CopyToClipboard text={gitRepo} onCopy={() => this.handleCopyToClipboard()}>
            <Button
              variant="link"
              icon={<CopyIcon />}
              name="Copy to Clipboard"
              data-testid="copy-to-clipboard"
            />
          </CopyToClipboard>
        </CheTooltip>
      </FormGroup>
    );
  }
}

export default GitRepoFormGroup;
