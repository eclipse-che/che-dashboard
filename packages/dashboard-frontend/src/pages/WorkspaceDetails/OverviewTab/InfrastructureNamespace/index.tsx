/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { FormGroup } from '@patternfly/react-core';
import React from 'react';

import overviewStyles from '@/pages/WorkspaceDetails/OverviewTab/index.module.css';

type Props = {
  namespace: string;
};

export class InfrastructureNamespaceFormGroup extends React.PureComponent<Props> {
  public render(): React.ReactElement {
    return (
      <FormGroup label="Kubernetes Namespace" fieldId="infrastructure-namespace">
        <div className={overviewStyles.readonly}>{this.props.namespace}</div>
      </FormGroup>
    );
  }
}
