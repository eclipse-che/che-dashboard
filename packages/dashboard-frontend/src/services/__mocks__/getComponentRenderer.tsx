/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

export default function <T extends Array<unknown>>(
  getComponent: (..._args: T) => React.ReactElement,
): {
  createSnapshot: (...args: T) => renderer.ReactTestRenderer;
  renderComponent: (...args: T) => {
    reRenderComponent: (..._args: T) => void;
  };
} {
  return {
    createSnapshot: (...args: T) => {
      return renderer.create(getComponent(...args));
    },
    renderComponent: (...args: T) => {
      const res = render(getComponent(...args));
      return {
        reRenderComponent: (..._args: T) => {
          res.rerender(getComponent(..._args));
        },
      };
    },
  };
}
