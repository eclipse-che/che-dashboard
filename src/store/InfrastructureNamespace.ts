/*
 * Copyright (c) 2018-2020 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Action, Reducer } from 'redux';
import { container } from '../inversify.config';
import { CheWorkspaceClient } from '../services/workspace-client/cheWorkspaceClient';
import { AppThunk } from './';

const WorkspaceClient = container.get(CheWorkspaceClient);

export interface State {
  isLoading: boolean;
  namespaces: che.KubernetesNamespace[];
}

interface RequestNamespacesAction {
  type: 'REQUEST_NAMESPACES';
}

interface ReceiveNamespacesAction {
  type: 'RECEIVE_NAMESPACES';
  namespaces: che.KubernetesNamespace[];
}

type KnownAction = RequestNamespacesAction
  | ReceiveNamespacesAction;

export type ActionCreators = {
  requestNamespaces: () => AppThunk<KnownAction, Promise<Array<che.KubernetesNamespace>>>;
};

export const actionCreators: ActionCreators = {

  requestNamespaces: (): AppThunk<KnownAction, Promise<Array<che.KubernetesNamespace>>> => async (dispatch): Promise<Array<che.KubernetesNamespace>> => {
    dispatch({ type: 'REQUEST_NAMESPACES' });

    try {
      const namespaces = await WorkspaceClient.restApiClient.getKubernetesNamespace<Array<che.KubernetesNamespace>>();
      dispatch({ type: 'RECEIVE_NAMESPACES', namespaces });
      return namespaces;
    } catch (e) {
      throw new Error('Failed to request list of available kubernetes namespaces, \n' + e);
    }
  },

};

const unloadedState: State = {
  isLoading: false,
  namespaces: [],
};

export const reducer: Reducer<State> = (state: State | undefined, incomingAction: Action): State => {
  if (state === undefined) {
    return unloadedState;
  }

  const action = incomingAction as KnownAction;
  switch (action.type) {
    case 'REQUEST_NAMESPACES':
      return Object.assign({}, state, {
        isLoading: true,
      });
    case 'RECEIVE_NAMESPACES':
      return Object.assign({}, state, {
        namespaces: action.namespaces,
      });
    default:
      return state;
  }
};
