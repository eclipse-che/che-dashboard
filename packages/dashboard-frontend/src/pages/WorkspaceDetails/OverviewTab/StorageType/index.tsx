/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import {
  Alert,
  AlertVariant,
  Button,
  FormGroup,
  Modal,
  ModalVariant,
  Radio,
  Text,
  TextContent,
  TextVariants,
} from '@patternfly/react-core';
import { OutlinedQuestionCircleIcon, PencilAltIcon } from '@patternfly/react-icons';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import overviewStyles from '@/pages/WorkspaceDetails/OverviewTab/index.module.css';
import styles from '@/pages/WorkspaceDetails/OverviewTab/StorageType/index.module.css';
import { che } from '@/services/models';
import * as storageTypeService from '@/services/storageTypes';
import { RootState } from '@/store';
import { selectBranding } from '@/store/Branding/selectors';
import { selectPvcStrategy } from '@/store/ServerConfig/selectors';

export type Props = MappedProps & {
  readonly: boolean;
  storageType?: che.WorkspaceStorageType;
  parentStorageType?: che.WorkspaceStorageType;
  onSave: (storageType: che.WorkspaceStorageType) => void;
};
export type State = {
  isSelectorOpen?: boolean;
  selected?: che.WorkspaceStorageType;
  isInfoOpen?: boolean;
};

class StorageTypeFormGroup extends React.PureComponent<Props, State> {
  storageTypes: che.WorkspaceStorageType[] = [];
  options: string[] = [];
  preferredType: che.WorkspaceStorageType;

  constructor(props: Props) {
    super(props);

    this.state = {
      isSelectorOpen: false,
      isInfoOpen: false,
    };

    const availableTypes = storageTypeService.getAvailable();

    if (Array.isArray(availableTypes)) {
      this.storageTypes = availableTypes;
      this.storageTypes.forEach(type => this.options.push(storageTypeService.toTitle(type)));
    }
    const preferredType = this.props.preferredStorageType as che.WorkspaceStorageType;
    if (preferredType) {
      this.preferredType = preferredType;
    }
  }

  public componentDidUpdate(prevProps: Props): void {
    if (prevProps.storageType !== this.props.storageType) {
      const selected = this.props.storageType;
      this.setState({ selected });
    }
  }

  public componentDidMount(): void {
    const selected = this.getSelection();
    this.setState({ selected });
  }

  private handleEditToggle(isSelectorOpen: boolean): void {
    this.setState({ isSelectorOpen });
  }

  private handleInfoToggle(): void {
    this.setState(({ isInfoOpen }) => ({
      isInfoOpen: !isInfoOpen,
    }));
  }

  private getExistingTypes(): {
    hasEphemeral: boolean;
    hasPerUser: boolean;
    hasPerWorkspace: boolean;
  } {
    const hasEphemeral = this.storageTypes.some(type => type === 'ephemeral');
    const hasPerUser = this.storageTypes.some(type => type === 'per-user');
    const hasPerWorkspace = this.storageTypes.some(type => type === 'per-workspace');

    return { hasEphemeral, hasPerUser, hasPerWorkspace };
  }

  private getInfoModalContent(): React.ReactNode {
    const { hasEphemeral, hasPerUser, hasPerWorkspace } = this.getExistingTypes();

    const ephemeralTypeDescr = hasEphemeral ? (
      <Text>
        <b>Ephemeral Storage</b> allows for faster I/O but may have limited storage and is not
        persistent.
      </Text>
    ) : (
      ''
    );
    const perUserTypeDescr = hasPerUser ? (
      <Text>
        <b>Per-user Storage</b> one PVC is provisioned per namespace. All of the workspace&apos;s
        storage (volume mounts) mounted in it on subpaths according to devworkspace ID.
      </Text>
    ) : (
      ''
    );
    const perWorkspaceTypeDescr = hasPerWorkspace ? (
      <Text>
        <b>Per-workspace Storage</b> a PVC is provisioned for each workspace within the namespace.
        All of the workspace&apos;s storage (volume mounts) are mounted on subpaths within the
        workspace&apos;s PVC.
      </Text>
    ) : (
      ''
    );

    const href = this.props.branding.docs.storageTypes;

    return (
      <TextContent>
        {perUserTypeDescr}
        {perWorkspaceTypeDescr}
        {ephemeralTypeDescr}
        <Text>
          <a rel="noreferrer" target="_blank" href={href}>
            Open documentation page
          </a>
        </Text>
      </TextContent>
    );
  }

  private getSelectorModal(): React.ReactNode {
    const { hasEphemeral, hasPerUser, hasPerWorkspace } = this.getExistingTypes();

    const isDisabled = this.props.parentStorageType !== undefined;
    const { isSelectorOpen, selected } = this.state;
    const originSelection = this.getSelection();

    const ephemeralTypeDescr = hasEphemeral ? (
      <Text component={TextVariants.h6}>
        <Radio
          label="Ephemeral"
          name="ephemeral"
          id="ephemeral-type-radio"
          description="Ephemeral Storage allows for faster I/O but may have limited storage and is not persistent."
          isChecked={selected === 'ephemeral'}
          isDisabled={isDisabled}
          onChange={() => this.setState({ selected: 'ephemeral' })}
        />
      </Text>
    ) : (
      ''
    );
    const perUserTypeDescr = hasPerUser ? (
      <Text component={TextVariants.h6}>
        <Radio
          label="Per-user"
          name="per-user"
          id="per-user-type-radio"
          description="Per-user Storage. One PVC is provisioned per user namespace and used by all workspaces of a given user"
          isChecked={selected === 'per-user'}
          isDisabled={isDisabled}
          onChange={() => this.setState({ selected: 'per-user' })}
        />
      </Text>
    ) : (
      ''
    );
    const perWorkspaceTypeDescr = hasPerWorkspace ? (
      <Text component={TextVariants.h6}>
        <Radio
          label="Per-workspace"
          name="per-workspace"
          id="per-workspace-type-radio"
          description="Per-workspace Storage. One PVC is provisioned for each workspace within the namespace."
          isChecked={selected === 'per-workspace'}
          isDisabled={isDisabled}
          onChange={() => this.setState({ selected: 'per-workspace' })}
        />
      </Text>
    ) : (
      ''
    );

    return (
      <Modal
        variant={ModalVariant.small}
        isOpen={isSelectorOpen}
        className={styles.modalEditStorageType}
        title="Change Storage Type"
        onClose={() => this.handleCancelChanges()}
        actions={[
          <Button
            key="confirm"
            variant="primary"
            isDisabled={originSelection === selected}
            onClick={() => this.handleConfirmChanges()}
          >
            Save
          </Button>,
          <Button key="cancel" variant="secondary" onClick={() => this.handleCancelChanges()}>
            Cancel
          </Button>,
        ]}
      >
        <TextContent>
          <Alert
            variant={AlertVariant.warning}
            className={styles.warningAlert}
            title={
              isDisabled
                ? 'Storage type is already defined in parent, you cannot change it.'
                : 'Note that after changing the storage type you may lose workspace data.'
            }
            isInline
          />
          <Text component={TextVariants.h6}>Select the storage type</Text>
          {perUserTypeDescr}
          {perWorkspaceTypeDescr}
          {ephemeralTypeDescr}
        </TextContent>
      </Modal>
    );
  }

  private handleConfirmChanges(): void {
    const selection = this.state.selected as che.WorkspaceStorageType;
    this.props.onSave(selection);
    this.setState({
      selected: selection,
      isSelectorOpen: false,
    });
  }

  private getSelection(): che.WorkspaceStorageType {
    if (this.props.parentStorageType) {
      return this.props.parentStorageType;
    } else if (this.props.storageType) {
      return this.props.storageType;
    }

    return this.preferredType;
  }

  private handleCancelChanges(): void {
    const originSelection = this.getSelection();
    this.setState({ selected: originSelection });
    this.handleEditToggle(false);
  }

  public render(): React.ReactNode {
    const { readonly } = this.props;
    const { selected, isInfoOpen } = this.state;

    return (
      <FormGroup
        label="Storage Type"
        fieldId="storage-type"
        labelIcon={
          <Button
            variant="plain"
            onClick={() => this.handleInfoToggle()}
            className={styles.labelIcon}
            title="Storage Type Info"
          >
            <OutlinedQuestionCircleIcon />
          </Button>
        }
      >
        {readonly && <span className={overviewStyles.readonly}>{selected}</span>}
        {!readonly && (
          <span className={overviewStyles.editable}>
            {selected}
            <Button
              data-testid="overview-storage-edit-toggle"
              variant="plain"
              onClick={() => this.handleEditToggle(true)}
              title="Change Storage Type"
            >
              <PencilAltIcon />
            </Button>
          </span>
        )}
        {this.getSelectorModal()}
        <Modal
          title="Storage Type Info"
          variant={ModalVariant.small}
          isOpen={isInfoOpen}
          onClose={() => {
            this.handleInfoToggle();
          }}
        >
          {this.getInfoModalContent()}
        </Modal>
      </FormGroup>
    );
  }
}

const mapStateToProps = (state: RootState) => ({
  branding: selectBranding(state),
  preferredStorageType: selectPvcStrategy(state),
});

const connector = connect(mapStateToProps, null, null, {
  // forwardRef is mandatory for using `@react-mock/state` in unit tests
  forwardRef: true,
});

type MappedProps = ConnectedProps<typeof connector>;
export default connector(StorageTypeFormGroup);
