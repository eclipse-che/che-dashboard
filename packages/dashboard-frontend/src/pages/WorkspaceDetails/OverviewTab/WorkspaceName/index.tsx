/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { ApplicationId } from '@eclipse-che/common';
import { Button, FormGroup } from '@patternfly/react-core';
import { ExternalLinkSquareAltIcon } from '@patternfly/react-icons';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import overviewStyles from '@/pages/WorkspaceDetails/OverviewTab/index.module.css';
import { Workspace, WorkspaceAdapter } from '@/services/workspace-adapter';
import { AppState } from '@/store';
import { actionCreators } from '@/store/BannerAlert';
import { selectApplications } from '@/store/ClusterInfo/selectors';

export type Props = MappedProps & {
  workspace: Workspace;
};

class WorkspaceNameFormGroup extends React.PureComponent<Props> {
  private buildOpenShiftConsoleLink(): React.ReactElement | undefined {
    const { applications, workspace } = this.props;
    const clusterConsole = applications.find(app => app.id === ApplicationId.CLUSTER_CONSOLE);

    if (!clusterConsole) {
      return;
    }

    const devWorkspaceOpenShiftConsoleUrl = WorkspaceAdapter.buildClusterConsoleUrl(
      workspace.ref,
      clusterConsole.url,
    );

    return (
      <Button
        component="a"
        variant="link"
        href={devWorkspaceOpenShiftConsoleUrl}
        target="_blank"
        icon={<ExternalLinkSquareAltIcon />}
        iconPosition="right"
        isInline
      >
        {workspace.name}
      </Button>
    );
  }

  public render(): React.ReactNode {
    const workspaceName = this.buildOpenShiftConsoleLink() || this.props.workspace.name;
    return (
      <FormGroup label="Workspace">
        <span className={overviewStyles.readonly}>{workspaceName}</span>
      </FormGroup>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  applications: selectApplications(state),
});

const connector = connect(mapStateToProps, actionCreators);

type MappedProps = ConnectedProps<typeof connector>;

export default connector(WorkspaceNameFormGroup);
