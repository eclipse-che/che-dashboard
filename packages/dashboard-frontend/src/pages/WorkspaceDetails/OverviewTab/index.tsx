/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { Form, PageSection, PageSectionVariants } from '@patternfly/react-core';
import StorageTypeFormGroup from './StorageType';
import { WorkspaceNameFormGroup } from './WorkspaceName';
import InfrastructureNamespaceFormGroup from './InfrastructureNamespace';
import ProjectsFormGroup from './Projects';
import { constructWorkspace, Workspace } from '../../../services/workspace-adapter';
import { cloneDeep } from 'lodash';
import { DevWorkspaceStatus } from '../../../services/helpers/types';
import { StorageTypeTitle } from '../../../services/storageTypes';

type Props = {
  onSave: (workspace: Workspace) => Promise<void>;
  workspace: Workspace;
};

export type State = {
  storageType: che.WorkspaceStorageType;
  workspaceName: string;
};

export class OverviewTab extends React.Component<Props, State> {
  private readonly workspaceNameCallbacks: { cancelChanges?: () => void };
  private isWorkspaceNameChanged: boolean;

  constructor(props: Props) {
    super(props);

    const { workspace } = this.props;

    this.state = {
      storageType: workspace.storageType,
      workspaceName: workspace.name,
    };

    this.isWorkspaceNameChanged = false;
    this.workspaceNameCallbacks = {};
  }

  public componentDidUpdate(): void {
    const { storageType, workspaceName } = this.state;
    const workspace = this.props.workspace;

    if (storageType !== workspace.storageType || workspaceName !== workspace.name) {
      this.setState({
        storageType: workspace.storageType,
        workspaceName: workspace.name,
      });
    }
  }

  public get hasChanges() {
    return this.isWorkspaceNameChanged;
  }

  public cancelChanges(): void {
    if (this.workspaceNameCallbacks.cancelChanges) {
      this.workspaceNameCallbacks.cancelChanges();
    }
  }

  private async handleWorkspaceNameSave(workspaceName: string): Promise<void> {
    if (this.props.workspace.isDevWorkspace) {
      return;
    }
    const workspaceClone = constructWorkspace(cloneDeep(this.props.workspace.ref));
    workspaceClone.name = workspaceName;
    await this.props.onSave(workspaceClone);
    this.setState({ workspaceName });
  }

  private async handleStorageSave(storageType: che.WorkspaceStorageType): Promise<void> {
    const workspaceClone = constructWorkspace(cloneDeep(this.props.workspace.ref));
    workspaceClone.storageType = storageType;
    await this.props.onSave(workspaceClone);
    this.setState({ storageType });
  }

  public render(): React.ReactElement {
    const { workspaceName, storageType } = this.state;
    const { workspace } = this.props;
    const namespace = workspace.namespace;
    const projects = workspace.projects;
    const isDeprecated = workspace.isDeprecated;

    return (
      <React.Fragment>
        <PageSection variant={PageSectionVariants.light}>
          <Form isHorizontal onSubmit={e => e.preventDefault()}>
            <WorkspaceNameFormGroup
              name={workspaceName}
              readonly={isDeprecated || workspace.isDevWorkspace}
              onSave={_workspaceName => this.handleWorkspaceNameSave(_workspaceName)}
              onChange={_workspaceName => {
                this.isWorkspaceNameChanged = workspaceName !== _workspaceName;
              }}
              callbacks={this.workspaceNameCallbacks}
            />
            <InfrastructureNamespaceFormGroup namespace={namespace} />
            <StorageTypeFormGroup
              readonly={isDeprecated || workspace.status === DevWorkspaceStatus.TERMINATING}
              storageType={StorageTypeTitle[storageType as che.WorkspaceStorageType] || storageType}
              onSave={_storageType => this.handleStorageSave(_storageType)}
            />
            <ProjectsFormGroup projects={projects} />
          </Form>
        </PageSection>
      </React.Fragment>
    );
  }
}

export default OverviewTab;
