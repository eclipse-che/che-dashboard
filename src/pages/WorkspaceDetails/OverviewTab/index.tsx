/*
 * Copyright (c) 2018-2020 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { Form, FormGroup, PageSection, PageSectionVariants } from '@patternfly/react-core';
import StorageTypeFormGroup from './StorageType';
import { WorkspaceNameFormGroup } from './WorkspaceName';
import InfrastructureNamespaceFormGroup from './InfrastructureNamespace';
import { updateDevfile } from '../../../services/storageTypes';

type Props = {
  onSave: (workspace: che.Workspace) => Promise<void>;
  workspace: che.Workspace;
};

export type State = {
  storageType: che.WorkspaceStorageType;
  namespace: string;
  workspaceName: string;
};

export class OverviewTab extends React.Component<Props, State> {
  private isWorkspaceNameChanged = false;
  private workspaceNameCallbacks: { cancelChanges?: () => void } = {};

  constructor(props: Props) {
    super(props);

    const { workspace } = this.props;
    const storageType = this.getStorageType(workspace.devfile);
    const workspaceName = workspace.devfile.metadata.name ? workspace.devfile.metadata.name : '';
    const namespace = workspace.attributes && workspace.attributes.infrastructureNamespace ? workspace.attributes.infrastructureNamespace : '';

    this.state = { storageType, workspaceName, namespace };
  }

  public get hasChanges() {
    return this.isWorkspaceNameChanged;
  }

  public cancelChanges(): void {
    if (this.workspaceNameCallbacks.cancelChanges) {
      this.workspaceNameCallbacks.cancelChanges();
    }
  }

  private async handleWorkspaceNameSave(workspaceName: string): Promise<void> {
    const newDevfile = Object.assign({}, this.props.workspace.devfile);
    newDevfile.metadata.name = workspaceName;
    this.setState({ workspaceName });
    await this.onSave(newDevfile);
  }

  private async handleStorageSave(storageType: che.WorkspaceStorageType): Promise<void> {
    const newDevfile = updateDevfile(this.props.workspace.devfile, storageType);
    this.setState({ storageType });
    await this.onSave(newDevfile);
  }

  private getStorageType(devfile: che.WorkspaceDevfile): che.WorkspaceStorageType {
    if (devfile.attributes && devfile.attributes.persistVolumes === 'false') {
      const isAsync = devfile.attributes && devfile.attributes.asyncPersist === 'true';
      if (isAsync) {
        return 'async';
      } else {
        return 'ephemeral';
      }
    } else {
      return 'persistent';
    }
  }

  public render(): React.ReactElement {
    const devfile = this.props.workspace.devfile;
    const storageType = this.getStorageType(devfile);
    const workspaceName = devfile.metadata.name ? devfile.metadata.name : '';
    const namespace = this.state.namespace;
    const projects = devfile.projects ? devfile.projects.map(project => project.name) : [];

    return (
      <React.Fragment>
        <PageSection
          variant={PageSectionVariants.light}
        >
          <Form isHorizontal onSubmit={e => e.preventDefault()}>
            <WorkspaceNameFormGroup
              name={workspaceName}
              onSave={_workspaceName => this.handleWorkspaceNameSave(_workspaceName)}
              onChange={_workspaceName => {
                this.isWorkspaceNameChanged = workspaceName !== _workspaceName;
              }}
              callbacks={this.workspaceNameCallbacks}
            />
            <InfrastructureNamespaceFormGroup namespace={namespace} />
            <StorageTypeFormGroup
              storageType={storageType}
              onSave={_storageType => this.handleStorageSave(_storageType)}
            />
            <FormGroup label="Projects" fieldId="projects">
              <div style={{ paddingTop: '5px', lineHeight: '30px' }}>{projects.join(', ')}</div>
            </FormGroup>
          </Form>
        </PageSection>
      </React.Fragment>
    );
  }

  private async onSave(devfile: che.WorkspaceDevfile): Promise<void> {
    const newWorkspaceObj = Object.assign({}, this.props.workspace);
    newWorkspaceObj.devfile = devfile;

    await this.props.onSave(newWorkspaceObj);
  }

}

export default OverviewTab;
