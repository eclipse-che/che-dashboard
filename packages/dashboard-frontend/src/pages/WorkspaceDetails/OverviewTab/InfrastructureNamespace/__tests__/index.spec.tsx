/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { InfrastructureNamespaceFormGroup } from '@/pages/WorkspaceDetails/OverviewTab/InfrastructureNamespace';
import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

describe('InfrastructureNamespaceFormGroup', () => {
  test('screenshot', () => {
    const snapshot = createSnapshot('user-namespace');
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test('namespace is displayed', () => {
    renderComponent('user-namespace');
    expect(screen.queryByText('user-namespace')).not.toBeNull();
  });
});

function getComponent(namespace: string) {
  return <InfrastructureNamespaceFormGroup namespace={namespace} />;
}
