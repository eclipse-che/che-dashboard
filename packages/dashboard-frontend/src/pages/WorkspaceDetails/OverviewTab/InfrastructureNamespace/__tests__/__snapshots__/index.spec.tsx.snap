// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfrastructureNamespaceFormGroup screenshot 1`] = `
<div
  className="pf-c-form__group"
>
  <div
    className="pf-c-form__group-label"
  >
    <label
      className="pf-c-form__label"
      htmlFor="infrastructure-namespace"
    >
      <span
        className="pf-c-form__label-text"
      >
        Kubernetes Namespace
      </span>
    </label>
     
  </div>
  <div
    className="pf-c-form__group-control"
  >
    <div
      className="readonly"
    >
      user-namespace
    </div>
  </div>
</div>
`;
